### ISSUE TITLE
False positives and false negatives with unused_optional_binding

### ISSUE THREAD

**Issue by drodriguez** [realm/SwiftLint#1376]
```
// Should not trigger, but triggers
if case .some(_) = self {
}

// Should trigger, but doesn’t
if case .some(let _) = self {
}
```

Seems that any usage of `case` is ignored.

Reviewing the rule code, the regular expressions used are quite broad and don’t even check for `let` being or not there.

Something like the following almost get it right:

```
        let underlineOutsideParenthesis = "(?<=let\\s)_\\s*="
        let underlineInsideParenthesis = "(?<=let\\s)\\((\\s*[_,]\\s*)+\\)"
        let pattern = "(("  + underlineOutsideParenthesis + ")|(" + underlineInsideParenthesis + "))"
```

But it doesn’t catch the `case .some(let _)` case.

/cc @marcelofabri

**Comment by crayment** [comment]
Getting false positive a lot for code like:

```
guard case .dictionary(_) = json else {
```

**Comment by marcelofabri** [comment]
@drodriguez Can you send a PR fixing the false positives? I think that is more important than catching the `case .some(let _)` case for now.

We can then create another issue just for that and address in the future.

**Comment by drodriguez** [comment]
PR sent with the full solution (hopefully): #1396

**Comment by aleph7** [comment]
I'm getting a false positive with 
```
if let point = state.find({ _ in true }) { ... }
```

**Comment by drodriguez** [comment]
I think the fix in #1396 will avoid the false positive in @aleph7 example. When I have a moment I will add it to the examples of that PR and check for sure.
